File: Study5 - Tree/Week32/Day220/boj16437_hsk.js
const fs = require('fs');
const stdin = (
  process.platform === 'linux'
    ? fs.readFileSync('/dev/stdin').toString().trim()
    : `4
S 100 3
W 50 1
S 10 1`
).split('\n');

const input = (() => {
  let line = 0;
  return () => stdin[line++];
})();

const getNumOfRescuedSheeps = (curNode) => {
  let numOfRescuedSheeps = Number(treeOfIslands[curNode].cntAnimals);

  for (let nextIsland of treeOfIslands[curNode].connectedIsland) {
    numOfRescuedSheeps += getNumOfRescuedSheeps(nextIsland);
  }

  if (numOfRescuedSheeps < 0) numOfRescuedSheeps = 0;

  return numOfRescuedSheeps;
};

const numOfIslands = parseInt(input());
const treeOfIslands = Array.from(new Array(numOfIslands + 1), () => new Object({ cntAnimals: 0, connectedIsland: [] }));
for (let i = 2; i <= numOfIslands; i++) {
  const [isWolf, numOfAnimals, island] = input().split(' ');

  treeOfIslands[i].cntAnimals = isWolf === 'W' ? -numOfAnimals : numOfAnimals;
  treeOfIslands[island].connectedIsland.push(i);
}

console.log(getNumOfRescuedSheeps(1));

File: Study5 - Tree/Week32/Day220/boj16437_hsk.py
# nodejs로 stack size exceeded 문제로 실패하여 python으로 풀이
import sys
input = sys.stdin.readline
sys.setrecursionlimit(10 ** 9)

class TreeOfIslands:
    def __init__(self):
        self.cntAnimals = 0
        self.connectedIsland = []

## 깔끔해요
## 이제 재귀 활용이 꽤나 자연스러워진 것 같아요.
def getNumOfRescuedSheeps(curNode):
    numOfRescuedSheeps = int(treeOfIslands[curNode].cntAnimals)

    for nextIsland in treeOfIslands[curNode].connectedIsland:
        numOfRescuedSheeps += getNumOfRescuedSheeps(nextIsland)

    if numOfRescuedSheeps < 0: numOfRescuedSheeps = 0
    
    return numOfRescuedSheeps


numOfIslands = int(input())
treeOfIslands = [TreeOfIslands() for _ in range(numOfIslands + 1)]

for i in range(2,numOfIslands + 1 ):
    isWolf, numOfAnimals, island = input().split(' ')

    ## 처음부터 음수 설정! 생각지 못 한 방법이네요. 다른 방법도 알아갑니다.
    treeOfIslands[i].cntAnimals = (-int(numOfAnimals) if isWolf == 'W' else int(numOfAnimals))
    treeOfIslands[int(island)].connectedIsland.append(i)

print(getNumOfRescuedSheeps(1))